### PR TITLE
Fixes issue #1957

### LINKED ISSUE #1957: Extension-generation needs adjustements
The extension-generation process is currently assuming you need encrypted userdata. That should get adjusted like this:
- Add an option to the click annotations like `--encrypted-userdata` here: https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/cli/cli_ext.py#L39 (including adding it as a param to the method signature)
- Add a section explaining what it means and that the extension should decide between encrypted/unencrypted here: https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/cli/cli_ext.py#L111-L112
- Add the variable to the `ExtGen` instantiation and there ...
- The service and the controller would needs some adjustments. They are rendered here: https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/cli/cli_ext.py#L150-L160 however, the code for the templates is living here:
  - https://github.com/cryptoadvance/specterext-dummy/blob/master/src/dummyorg/specterext/dummy/service.py#L12
  - https://github.com/cryptoadvance/specterext-dummy/blob/master/src/dummyorg/specterext/dummy/controller.py#L53
  
Mind the different templating syntax. Those are "metatemplates" as they are creating templates which themself are later templates.

You can test it via using `--tmpl-fs-source` and checkout the dummy-project somewhere. Also i've created a small section about https://docs.specter.solutions/desktop/extensions/intro/#virtualenv-management

### PR DESCRIPTION
**Pull request by Aditi-Singh16**
fixes #1957

### CODE REVIEW

**Comment by k9ert** [comment]
Great! However, the test is failing. You have to adjust it as you changed the constructor.

File: src/cryptoadvance/specter/services/extension_gen.py [hunk 1]
```diff
@@ -46,13 +48,20 @@ def __init__(
...
+        self.custom_version = custom_version
         self.isolated_client = isolated_client
         self.devicename = devicename
         self.author = author
         self.author_email = email
         vc = VersionChecker()
         version = vc._get_current_version()
-        if version == "custom":
+        if self.custom_version != "":
+            if vc._check_if_version_is_available(custom_version) == True:
```
Comment (k9ert) [inline src/cryptoadvance/specter/services/extension_gen.py:60]: `== True` is not needed. In order to improve readability, you could rename the method and call it:
```suggestion
            if vc._version_is_available(custom_version):
```
Comparing methods which return booleans explicitely is a common issue. It's not wrong but i hope the readability improvement above makes sense to you.

File: src/cryptoadvance/specter/services/extension_gen.py [hunk 2]
```diff
@@ -46,13 +48,20 @@ def __init__(
...
+        self.encrypted_userdata = encrypted_userdata
+        self.custom_version = custom_version
         self.isolated_client = isolated_client
         self.devicename = devicename
         self.author = author
         self.author_email = email
         vc = VersionChecker()
         version = vc._get_current_version()
-        if version == "custom":
+        if self.custom_version != "":
```
Comment (k9ert) [inline src/cryptoadvance/specter/services/extension_gen.py:59]: Please have a look how `branch` and `version` is used. the feature i mostly spoke about was using specific branches (tags) as the basis for the templating [here](https://github.com/cryptoadvance/specter-desktop/blob/master/src/cryptoadvance/specter/services/extension_gen.py#L66).
Yes, the `self.version` is used for specific templates in order to render out the version this extension is dependent on. But the branch variable specifies which templates are used: [branches (or tags, doesn't matter) in the specterext-dummy-repo](https://github.com/cryptoadvance/specterext-dummy/tags). You can rename "branch" to "tag" which would make more sense.

File: src/cryptoadvance/specter/cli/cli_ext.py [hunk 3]
```diff
@@ -109,6 +126,26 @@ def gen(org, ext_id, isolated_client, devicename, tmpl_fs_source, dryrun):
...
+            """
+        )
+        encrypted_userdata = click.prompt(
+            "Should the data be encrypted (y/n)?",
+            type=bool,
+        )
+    if custom_version == None:
+        print(
+            """
+            The extension generation uses the latest version of specter.
```
Comment (k9ert) [inline src/cryptoadvance/specter/cli/cli_ext.py:142]: I think a fallback should rather be not the lastest version, but the version ext_gen is running on. That can be found (in <= v2) in `from cryptoadvance.specter._version import version`.

File: src/cryptoadvance/specter/util/version.py [hunk 4]
```diff
@@ -72,6 +72,37 @@ def _get_current_version(self):
             current = "custom"
         return current
 
+    def _check_if_version_is_available(self, custom_version):
+        """
+        checks if version requested is available or not.
```
Comment (k9ert) [inline src/cryptoadvance/specter/util/version.py:77]:
```suggestion
        checks if that version is available as a package on pypi.
```

**Comment by k9ert** [comment]
Looks You haven't addressed the test yet. Please run the tests and fix them. If you click on "details" and then on "Raw output", you see the reason why the test is failing:
```
caplog = <_pytest.logging.LogCaptureFixture object at 0x7f4b0a34fdf0>

    def test_ExtGen(caplog):
    
        with patch("cryptoadvance.specter.services.extension_gen.GithubUrlLoader") as mock:
>           extgen = ExtGen(
                ".",
                "testorg",
                "testext",
                False,
                False,
                "Some Author",
                "some@mail",
                # uncomment the below line to see a real generation
                # tmpl_fs_source="../specterext-dummy",
                dry_run=True,
            )
E           TypeError: ExtGen.__init__() missing 2 required positional arguments: 'author' and 'email'

tests/test_services_extension_gen.py:14: TypeError
```

**Review by k9ert — CHANGES_REQUESTED** [review]
see the above comments

**Comment by k9ert** [comment]
BTW, most important from all: This PR is going to master but should go to uiux-revamp branch.

**Comment by k9ert** [comment]
Sorry, we have moved the uiux-revamp branch to the master branch which now results with a lot of conflicts. So this PR needs to be recreated. Please do something like this:
```
git checkout master
git pull upstream master
git checkout -b ext-update-new
git cherry-pick 5814c687dd555e92fc7315c1749eb4fa9acf510a 1c7436b51a14794ffc637dab41bbcaf71e86ff1c cc706a0b7e20681ed6cec9b8549ae106e1e2e3a7 fc831ba86ea93d22b2328eeaaa45cfd5be2c5f4c
git commit
git push
```

And then please recreate the PR.

**Comment by Aditi-Singh16** [comment]
Okay sure!

**Pull request closed without merging**